include/alp/algorithms/backsubstitution.hpp

```diff
@@ -0,0 +1,131 @@
+
+	namespace algorithms {
+
+		/**
+		 * @brief Solves linear system Ax=b
+		 *        where A is UpperTriangular matrix, b is given RHS vector
+		 *        and x is the solution.
+		 *
+		 * @tparam D        Data element type
+		 * @tparam Ring     Type of the semiring used in the computation
+		 * @tparam Minus    Type minus operator used in the computation
+		 * @tparam Divide   Type of divide operator used in the computation
+		 * @param[in]  A    input upper trinagular matrix
+		 * @param[in]  b    input RHS vector
+		 * @param[out] x    solution vector
+		 * @param[in]  ring The semiring used in the computation
+		 * @return RC       SUCCESS if the execution was correct
+		 *
+		 */
+		template<
+			typename D = double,
+			typename Vecx,
+			typename Vecb,
+			typename Ring = Semiring< operators::add< D >, operators::mul< D >, identities::zero, identities::one >,
+			typename Minus = operators::subtract< D >,
+			typename Divide = operators::divide< D > >
+		RC backsubstitution(
+			Matrix< D, structures::UpperTriangular, Dense > &A,
+			Vecx &x,
+			Vecb &b,
+			const Ring &ring = Ring(),
+			const Minus &minus = Minus(),
+			const Divide &divide = Divide()
+		) {
+
+			RC rc = SUCCESS;
+
+			if( ( nrows( A ) != size( x ) ) || ( size( b ) != size( x ) ) ) {
+				std::cerr << "Incompatible structures in trsv.\n";
+				return FAILED;
+			}
+
+			const size_t n = nrows( A );
+
+			for( size_t k = 0; k < n ; ++k ) {
+				Scalar< D > alpha( ring.template getZero< D >() );
+				const size_t i = n - k - 1;
+				//x[i]=(b[i]-A[i,i:].dot(x[i:]))/A[i,i]
+ 				auto A_i  = get_view( A, i, utils::range( i, n ) );
+				auto A_ii  = get_view( A, i, utils::range( i, i + 1 ) );
+				auto x_i  = get_view( x, utils::range( i, i + 1 ) );
+				auto b_i  = get_view( b, utils::range( i, i + 1 ) );
+				auto x_i_n  = get_view( x, utils::range( i, n ) );
+				rc = rc ? rc : alp::dot( alpha, A_i, alp::conjugate( x_i_n ), ring );
+				rc = rc ? rc : alp::set( x_i, b_i );
+				rc = rc ? rc : alp::foldl( x_i, alpha, minus );
+ 				rc = rc ? rc : alp::set( alpha, Scalar< D >( ring.template getZero< D >() ) );
+ 				rc = rc ? rc : alp::foldl( alpha, A_ii, ring.getAdditiveMonoid() );
+ 				rc = rc ? rc : alp::foldl( x_i, alpha, divide );
+			}
+
+			assert( rc == SUCCESS );
+			return rc;
+		}
+
+		template<
+			typename D = double,
+			typename StructX,
+			typename StructB,
+			typename Ring = Semiring< operators::add< D >, operators::mul< D >, identities::zero, identities::one >,
+			typename Minus = operators::subtract< D >,
+			typename Divide = operators::divide< D > >
+		RC backsubstitution(
+			Matrix< D, structures::UpperTriangular, Dense > &A,
+			Matrix< D, StructX, Dense > &X,
+			Matrix< D, StructB, Dense > &B,
+			const Ring &ring = Ring()
+		) {
+
+			RC rc = SUCCESS;
+
+			if (
+				( nrows( X ) != nrows( B ) ) ||
+				( ncols( X ) != ncols( B ) ) ||
+				( ncols( A ) != nrows( X ) )
+			) {
+				std::cerr << "Incompatible structures in trsm.\n";
```

Comment:
Maybe "incompatible sizes" rather than structures, because that is what the if checks.